doc/ugidd.man
.TH UGIDD 8 "11 August 1997"
.SH NAME
ugidd \- uid/gid mapping daemons
.SH SYNOPSIS
.nf
.B /usr/sbin/rpc.ugidd
.\".LP
.\"or from SunOS inetd(8):
.\".B rpc udp /usr/etc/rpc.ugidd 545580417 1
.fi
.SH DESCRIPTION
.IX  "uid/gid mapping daemons"  "ugidd daemon"  ""  "\fLugidd\fP daemon"
.I ugidd
starts a
daemon that handles rpc requests to map uid/gids to string names and vice versa.
It is called by the \fInfsd\fP(8) server when the client and server do
not share the same passwd file.
.P
This version allows you to restrict access to the server through the
\fIhosts_access\fP(5) files when compiled with the -DENABLE_HOSTS_ACCESS option.
Otherwise, \fIugidd\fP could be used by anyone in the Internet to obtain
a list of all valid usernames on the client machine. For instance, to 
restrict access to hosts on your class C IP network, you could add the 
following lines to \fIhosts.allow\fP and \fIhosts.deny\fP:
.PP
.nf
.I "# /etc/hosts.allow
.I "rpc.ugidd:     192.168.1.
.I "# /etc/hosts.deny
.I "rpc.ugidd:     ALL
.fi
.SH "SEE ALSO"
exports(5), inetd(8), hosts_access(5)
.SH BUGS
The rpc program number is legal in that it lies in the 0x20000000
to 0x3fffffff ``Defined by user'' range, but is completely arbitrary.


ChangeLog
Tue Sep  6 13:13:04 2005 Sam Robb <sam.robb@timesys.com>

	* Continued cleanup of splint warnings.  This touches just about
	  every source file in the project, so I'm not even going to try
	  to list each one individually. 

Tue Aug  9 04:13:57 2005 Sam Robb <sam.robb@timesys.com>

	* BUGS : copied to BUGS.HISTORICAL, edited to reflect new maintainer
	  email address

	* Makefiles: added targets for static analysis.  Touched just about every
	  source and Makefile in the process.  Performing an anlysis with lowest
	  level of warnings (+weak) enabled in splint, there are > 200 errors
	  *after* significant cleanup.  Oy.

	* libnfs/fhandle.c (fh_buildpath) : incorporate patch from Linus Hicks to
	  fix 'Cannot stat <file>' error when ascending directory hierarchy when
	  a file is not found.

	* libnfs/fhandle.c (path_psi) : incorporated patch from Linus Hicks to
	  change path_psi inode handling to match that in fh_buildpath.

	* libnfs/fhandle.c : remove usage of FHTRACE

	* libnfs/fhandle.c (fh_verify_cache) : added

	* libnfs/fhandle.c (fh_flush) : call fh_verify_cache() after cache flush
	  if DEBUG is defined

	* libnfs/logging.c (log_enable) : added 'most' target for logging, to allow
	  a simple method for enabling all logging except user/group ID translation
	  messages (which are extremely prolific...)

	* nfsd/getattr.c (fhc_getattr) : return st_size in st_rdev field of stat
	  struct if HAVE_STRUCT_STAT_ST_RDEV is not defined

	* nfsd/ugid_map.c (ugid_get_entry) : consolidate multiple log lines into a
	  single log entry

Fri Jul 29 17:00:00 2005 Sam Robb <sam.robb@timesys.com>

	* configure.in : add check for stat structure member st_rdev

	* aclocal.m4 : regenerated

	* configure : regenerated

	* include/system.h : remove transparent redefinition of st_rdev as st_size

	* nfsd/getattr.c (st_rdev) : added

	* nfsd/getattr.c (fhc_getattr): conditionally call st_rdev() in instances
	  where st_rdev needs to be accessed, in order to properly munge st_rdev
	  to export linux device nodes from cygwin

	* nfsd/nfsd.c : use HAVE_STRUCT_STAT_ST_RDEV to determine whether or not
	  to place st_rdev or st_size into the returned stat info for a file

	* CYGWIN-PATCHES/nfs-server-config : Change static_map to map_static in
	  comments within auto-generated map files

Wed Apr  6 17:47:20 2005 Sam Robb <sam.robb@timesys.com>

	* nfsd/nfsd.c (nfsd_nfsproc_readdir_2) : remove workaround for cygwin
	  readdir problem.  Now dependent on cygwin 1.5.14 or better.

	* nfs-server-config : made more interctive, allow user to install
	  under system account or another account, ask user if existing
	  services should be reinstalled.

Thu Mar 31 16:22:03 2005 Sam Robb <sam.robb@timesys.com>

	* libnfs/fhandle.c : more descriptive fd cache inconsistency messages
	
	* doc/nfsd.man : document specifying root uid on command line 

	* include/auth.h : make ROOT_UID a variable instead of a constant
	
	* include/libnfs/auth_clnt.c : use root_uid instead of ROOT_UID
	
	* include/libnfs/devtab.c : ditto

	* include/libnfs/fhandle.c : ditto

	* include/nfsd/dispatch.c : ditto

	* include/nfsd/nfsd.c : ditto

	* nfs-server.lsm : removed 

	* nfs-server-config : install nfsd with --root-uid argument specifying
	  the id of the user executing the script; generate default server map
	  based on executing user, Administrator uids/gids
	
	* server.map.example : added
	  
Thu Mar 17 16:51:40 2005 Sam Robb <sam.robb@timesys.com>

	* libnfs/logging.c (log_call) : add file, line no. parameters

	* mountd/dispatch.c (mount_dispatch) : ditto

	* nfsd/dispatch.c (nfs_dispatch) : ditto

	* nfsd/nfsd.c (nfsd_nfsproc_readdir_2) : temporary work around for
	  problem with cygwin readdir implementation.

Thu Mar 10 13:47:20 2005 Sam Robb <sam.robb@timesys.com>

	* configure : regenerated.
	
	* libnfs/auth.c (hostmatch) : allow '*' for hostname in exports file
	
Tue Dec  2 15:56:42 2003 Sam Robb <sam.robb@timesys.com>

	* Makefile.in : use mandir instead of hardcoded paths.

Fri Oct 24 16:24:06 2003 Jeremiah Lott <jeremiah.lott@timesys.com>

	* libnfs/fhandle.c : Change inode code to handle CYGWIN 64
	  bit inode better.  Also, made psuedo inode type 64 bits,
	  and made the pseudo_inode function a noop (identity function).

	* include/fhandle.h : Ditto.

Tue Oct  7 10:17:20 2003 Sam Robb <sam.robb@timesys.com>

	* nfs-server-config : incorporate suggested improvements
	  from Jari Aalto.

Thu Feb 13 11:36:23 2003 Sam Robb <sam.robb@timesys.com>

	* ilibnfs/auth_clnt.c (seteids) : do not need to change
	  to root in order to seteuid() under Cygwin

Tue Feb 11 17:16:51 2003 Sam Robb <sam.robb@timesys.com>

	* mountd/rmtab.c (rmtab_gethost) : return dotted quad
	  address if unable to resolve the hostname.

Tue Feb 11 14:36:55 2003 Sam Robb <sam.robb@timesys.com>

	* include/xrealpath.h : added

	* libnfs/xrealpath.c : added

	* libnfs/Makefile.in : added xrealpath.o

	* libnfs/auth_init.c (auth_init) : use xrealpath

	* mountd/mountd.c (mountproc_mnt_1) : use xrealpath

	* mountd/mountd.c (mountproc_pathconf_1) : use xrealpath

Fri Feb  7 14:20:38 2003 Sam Robb <sam.robb@timesys.com>

	* libnfs/fhandle.c : revisited the way that fh_buildpath()
	  corrects for differences between readdir() and stat()
	  under Cygwin.

Thu Jan 16 01:17:19 2003 Sam Robb <sam.robb@timesys.com>

	* include/fhandle.h : changed FH_CACHE_LIMIT to 2048; added
	  new define FH_CACHE_FLUSH_RATIO

	* libnfs/fhandle.c (fh_flush) : rewrote to try and avoid work
	  whenever possible

Tue Jan 14 12:48:36 2003 Sam Robb <sam.robb@timesys.com>

	* libnfs/auth.c (auth_init_lists) : under Cygwin, anonymous
	  uid/gid is that of user "nobody", or that of user "Guest",
	  or -2/-2 if netiher of those two accounts exits

	* doc/exports.man : added section documenting behavior under
	  Cygwin when checking for anonymous user uid/gid

	* nfsd/nfsd.c (main) : changed to dump call profiling stats
	  on SIGUSR2 instead of SIGIOT


ugidd/ugidd.c

/* UNFSD - copyright Mark A Shand, May 1988.
 *
 * This software maybe be used for any purpose provided
 * the above copyright notice is retained.  It is supplied
 * as is, with no warranty expressed or implied.
 *
 * Authors:	Mark A. Shand
 *			Olaf Kirch <okir@monad.swb.de>
 */

#include "system.h"
#include <rpc/pmap_clnt.h>
#include <getopt.h>
#include "logging.h"
#include "haccess.h"
#include "ugid_xdr.c"

#ifndef HAVE_RPCGEN_C
#define authenticate_1_svc	authenticate_1
#define name_uid_1_svc		name_uid_1
#define group_gid_1_svc		group_gid_1
#define uid_name_1_svc		uid_name_1
#define gid_group_1_svc		gid_group_1
#endif

static void ugidprog_1(struct svc_req *rqstp, SVCXPRT * transp);
static void usage(void);

static struct option longopts[] = {
	{"debug", 0, 0, 'd'},
	{NULL, 0, 0, 0}
};

int
main(int argc, char **argv)
{
	SVCXPRT *transp;
	int c;
	int longind;
	int foreground = 0;

#ifndef ENABLE_HOSTS_ACCESS
	fprintf(stderr,
		"\n *** WARNING: This copy of ugidd has been compiled without\n"
		" *** support for host_access checking. This is very risky.\n"
		" *** Please consider recompiling it with access checking.\n");
	sleep(1);
#endif

	while ((c = getopt_long(argc, argv, "d", longopts, &longind)) != EOF) {
		switch (c) {
		case 'd':
			foreground = 1;
			log_enable("ugid");
			break;
		default:
			usage();
		}
	}

	pmap_unset(UGIDPROG, UGIDVERS);

	transp = svcudp_create(RPC_ANYSOCK);
	if (transp == NULL) {
		(void) fprintf(stderr, "cannot create udp service.\n");
		exit(1);
	}

	if (!svc_register
	    (transp, UGIDPROG, UGIDVERS, ugidprog_1, IPPROTO_UDP)) {
		fprintf(stderr,
			"unable to register (UGIDPROG, UGIDVERS, UDP)\n");
		exit(1);
	}

	transp = svctcp_create(RPC_ANYSOCK, 0, 0);
	if (transp == NULL) {
		fprintf(stderr, "cannot create tcp service.\n");
		exit(1);
	}

	if (!svc_register
	    (transp, UGIDPROG, UGIDVERS, ugidprog_1, IPPROTO_TCP)) {
		fprintf(stderr,
			"unable to register (UGIDPROG, UGIDVERS, TCP)\n");
		exit(1);
	}

	if (!foreground) {
		pid_t child;
		if ((child = fork()) > 0) {
			exit(0);
		}

		if (child < 0) {
			fprintf(stderr, "ugidd: cannot fork: %s\n",
				strerror(errno));
			exit(-1);
		}

		(void) close(0);
		(void) close(1);
		(void) close(2);

#ifdef HAVE_SETSID
		(void) setsid();
#else
		{
			int fd;

			if ((fd = open("/dev/tty", O_RDWR)) >= 0) {
				ioctl(fd, TIOCNOTTY, (char *) NULL);
				close(fd);
			}
		}
#endif /* HAVE_SETSID */
	}

	log_open("ugidd", foreground);

	svc_run();

	dbg_printf(__FILE__, __LINE__, L_ERROR, "svc_run() returned\n");

	return 1;
}

static void
usage()
{
	fprintf(stderr, "rpc.ugidd: [-d]\n");
	exit(2);
}

static void
ugidprog_1(struct svc_req *rqstp, SVCXPRT * transp)
{
	union {
		int authenticate_1_arg;
		ugname name_uid_1_arg;
		ugname group_gid_1_arg;
		int uid_name_1_arg;
		int gid_group_1_arg;
	} argument;

	char *result;
	xdrproc_t xdr_argument;
	xdrproc_t xdr_result;
	char *(*local) ();

	if (!client_checkaccess("rpc.ugidd", svc_getcaller(transp), 1)) {
		return;
	}

	switch (rqstp->rq_proc) {
	case NULLPROC:
		svc_sendreply(transp, (xdrproc_t) xdr_void, (char *) NULL);
		return;

	case AUTHENTICATE:
		xdr_argument = (xdrproc_t) xdr_int;
		xdr_result = (xdrproc_t) xdr_int;
		local = (char *(*)()) authenticate_1_svc;
		break;

	case NAME_UID:
		xdr_argument = (xdrproc_t) xdr_ugname;
		xdr_result = (xdrproc_t) xdr_int;
		local = (char *(*)()) name_uid_1_svc;
		break;

	case GROUP_GID:
		xdr_argument = (xdrproc_t) xdr_ugname;
		xdr_result = (xdrproc_t) xdr_int;
		local = (char *(*)()) group_gid_1_svc;
		break;

	case UID_NAME:
		xdr_argument = (xdrproc_t) xdr_int;
		xdr_result = (xdrproc_t) xdr_ugname;
		local = (char *(*)()) uid_name_1_svc;
		break;

	case GID_GROUP:
		xdr_argument = (xdrproc_t) xdr_int;
		xdr_result = (xdrproc_t) xdr_ugname;
		local = (char *(*)()) gid_group_1_svc;
		break;

	default:
		svcerr_noproc(transp);
		return;
	}

	memset(&argument, 0, sizeof(argument));

	if (!svc_getargs(transp, xdr_argument, &argument)) {
		svcerr_decode(transp);
		return;
	}

	result = (*local) (&argument, rqstp);

	if (result != NULL && !svc_sendreply(transp, xdr_result, result)) {
		svcerr_systemerr(transp);
	}

	if (!svc_freeargs(transp, xdr_argument, &argument)) {
		fprintf(stderr, "unable to free arguments\n");
		exit(1);
	}
}

int *
authenticate_1_svc(int *argp, struct svc_req *rqstp)
{
	static int res;
	int s = 0;
	struct sockaddr_in sendaddr;
	struct sockaddr_in destaddr;
	int dummy;
	short lport;

	if ((*argp < 0) || (*argp > 65535)) {
		goto bad;
	}

	memset(&res, 0, sizeof(res));
	destaddr = *svc_getcaller(rqstp->rq_xprt);
	destaddr.sin_port = htons((in_port_t) *argp);

	if ((s = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
		goto bad;
	}

	setsockopt(s, SOL_SOCKET, SO_LINGER, 0, 0);
	memset(&sendaddr, 0, sizeof(sendaddr));

	/* find a reserved port */
	lport = IPPORT_RESERVED - 1;
	sendaddr.sin_family = (sa_family_t) AF_INET;
	sendaddr.sin_addr.s_addr = INADDR_ANY;

	for (;;) {
		sendaddr.sin_port = htons((u_short) lport);

		if (bind(s, (struct sockaddr *) &sendaddr, (int) sizeof(sendaddr)) >=
		    0) {
			break;
		}

		if (errno != EADDRINUSE && EADDRNOTAVAIL) {
			goto bad;
		}

		lport--;

		if (lport <= IPPORT_RESERVED / 2) {
			/* give up */
			break;
		}
	}

	if (sendto(s, &dummy, sizeof dummy, 0,
		   (struct sockaddr *) &destaddr, (int) sizeof(destaddr)) < 0) {
		goto bad;
	}

	close(s);
	res = 0;
	return (&res);

      bad:
	close(s);
	res = errno == 0 ? -1 : errno;
	return (&res);
}

int *
name_uid_1_svc(ugname * argp, struct svc_req *rqstp)
{
	static int res;
	struct passwd *pw;

	memset(&res, 0, sizeof(res));

	if ((pw = getpwnam(*argp)) == NULL) {
		res = NOBODY;
	} else {
		res = (int) pw->pw_uid;
	}

	return (&res);
}

int *
group_gid_1_svc(ugname * argp, struct svc_req *rqstp)
{
	static int res;
	struct group *gr;

	memset(&res, 0, sizeof(res));

	if ((gr = getgrnam(*argp)) == NULL) {
		res = NOBODY;
	} else {
		res = (int) gr->gr_gid;
	}

	return (&res);
}

ugname *
uid_name_1_svc(int *argp, struct svc_req * rqstp)
{
	static ugname res;
	struct passwd *pw;

	memset(&res, 0, sizeof(res));

        if ((*argp < 0) || (*argp > (int) ((gid_t) -1))) {   
                res = ""; 
                return (&res);
        } 

	if ((pw = getpwuid((uid_t) *argp)) == NULL) {
		res = "";
	} else {
		res = pw->pw_name;
	}

	return (&res);
}

ugname *
gid_group_1_svc(int *argp, struct svc_req * rqstp)
{
	static ugname res;
	struct group *gr;

	memset(&res, 0, sizeof(res));

	if ((*argp < 0) || (*argp > (int) ((gid_t) -1))) {
		res = "";
		return (&res);
	}

	if ((gr = getgrgid((gid_t) *argp)) == NULL) {
		res = "";
	} else {
		res = gr->gr_name;
	}

	return (&res);
}
